src/deluge/gui/views/session_view.cpp

```diff
@@ -4278,41 +4296,263 @@ char const* SessionView::getMacroKindString(SessionMacroKind kind) {
 }
 
 ActionResult SessionView::gridHandleScroll(int32_t offsetX, int32_t offsetY) {
-	if (currentUIMode == UI_MODE_CLIP_PRESSED_IN_SONG_VIEW && offsetY != 0) {
-		auto track = gridTrackFromX(gridFirstPressedX, gridTrackCount());
-		if (track != nullptr) {
-			if (Buttons::isButtonPressed(hid::button::Y_ENC)) {
+	if (currentUIMode == UI_MODE_CLIP_PRESSED_IN_SONG_VIEW
+	    && (Buttons::isShiftButtonPressed() || clipWasSelectedWithShift)) {
+		return gridChangeTrackColour(offsetX, offsetY);
+	}
+
+	// This is the right place to add new features like moving clips or tracks :)
+	bool movingClip = (currentUIMode == UI_MODE_CLIP_PRESSED_IN_SONG_VIEW && offsetY != 0);
+	bool movingTrack = (currentUIMode == UI_MODE_CLIP_PRESSED_IN_SONG_VIEW && offsetX != 0);
+
+	ActionResult result;
+
+	if (movingClip) {
+		result = gridHandleClipMove(offsetY);
+	}
+	else if (movingTrack) {
+		result = gridHandleTrackMove(offsetX);
+	}
+	else {
+		gridHandleVerticalScroll(offsetY);
+		gridHandleHorizontalScroll(offsetX);
+		result = ActionResult::DEALT_WITH;
+	}
+
+	// use root UI in case this is called from performance view
+	requestRendering(getRootUI(), 0xFFFFFFFF, 0xFFFFFFFF);
+	view.flashPlayEnable();
+
+	return result;
+}
+
+ActionResult SessionView::gridChangeTrackColour(int32_t offsetX, int32_t offsetY) {
+	auto track = gridTrackFromX(gridFirstPressedX, gridTrackCount());
+	if (track != nullptr) {
+		// this changes the track colour gradient (colours between each full colour)
+		if (offsetY != 0) {
+			track->colour += offsetY;
+			if (track->colour == 0) {
 				track->colour += offsetY;
-				if (track->colour == 0) {
-					track->colour += offsetY;
-				}
 			}
-			else {
-				track->colour = static_cast<int16_t>(track->colour + (colourStep * offsetY) + 192) % 192;
+		}
+		// this changes the track colour from one colour to the next
+		else if (offsetX != 0) {
+			track->colour = static_cast<int16_t>(track->colour + (colourStep * offsetX) + 192) % 192;
+		}
+		requestRendering(this);
+	}
+
+	return ActionResult::DEALT_WITH;
+}
+
+ActionResult SessionView::gridHandleClipMove(int32_t offsetY) {
```

Comment:
This is extremely difficult to read, and breaking the semantic meaning of the grid presses will lead to bugs (e.g. x/y no longer represent where a users finger is)